### PR TITLE
Remove Python 3.6

### PR DESCRIPTION
**Pull request by Kludex**
I don't think maintaining Python 3.6 is a burden for Starlette (I have a different opinion on Uvicorn tho). 

If we remove it on Uvicorn, should we also do it on Starlette?

This PR is to allow discussion having the changes that it implies in the same place. :)

Reference: https://pypistats.org/packages/starlette

Updated 14/01/22: `anyio` [3.4.0](https://anyio.readthedocs.io/en/stable/versionhistory.html) supports a feature to 3.7+.
Updated 27/03/22: `setuptools` and `jinja2` doesn't support Python 3.6 anymore.

Note to future me: remove `jinja2` import conditional about `context` function.

### CODE REVIEW

File: setup.py
Comment (Kludex) [inline setup.py]: ups! :sweat_smile: 

```suggestion
from setuptools import setup, find_packages
```

**Comment by Kludex** [comment]
Is that a visual bug on the pipeline? Python 3.6 is not running.

**Comment by aminalaee** [comment]
> Is that a visual bug on the pipeline? Python 3.6 is not running.

Isn't it expected since you removed it in your branch? This check won't run in your PR.
I guess it's the same when you add a new Matrix, in that PR the new matrix won't be required since the changes aren't in master.

**Comment by Kludex** [comment]
Ok... It makes sense... How do people merge PRs like this one, then? :thinking:

**Comment by aminalaee** [comment]
> Ok... It makes sense... How do people merge PRs like this one, then? 🤔

As far as I remember the pipeline check won't be required, the approval will be enough. Might need double checking...

**Comment by Kludex** [comment]
> > Ok... It makes sense... How do people merge PRs like this one, then? thinking
> 
> As far as I remember the pipeline check won't be required, the approval will be enough. Might need double checking...

Don't worry. I just open this for discussion initially. We'll see later. :)

**Review by aminalaee — DISMISSED** [review]

**Comment by Kludex** [comment]
Only @tomchristie can merge this, as the pipeline will never pass.

@tomchristie The reference about downloads per user is in the description of this PR, jfyk.

**Comment by lovelydinosaur** [comment]
I've updated the settings so that 3.7-3.10 are required.

I don't have any strong feelings on this one. I suppose the sensible thing is mostly just that we update our packages consistently when we go for this. (Not for ay hard reasons, just because at least they're all on the same page then.)

What is FastAPI's current version support?

Do we have an outstanding PR for this change in uvicorn too?

**Comment by aminalaee** [comment]
Sorry to jump in here, I don't know about FastAPI's plans for python 3.6 but this pypistats might be worth _considering_:
Screenshot 2022-01-10 at 10 46 20

Update: I'm not sure how accurate the numbers are but ignoring the recent peaks, there a few thousand downloads of Python3.6 happening every day.

**Comment by lovelydinosaur** [comment]
Link to those FastAPI stats... https://pypistats.org/packages/fastapi

**Comment by Kludex** [comment]
I've talked to @tiangolo about this, and we reached the conclusion that we can wait some time before losing support for Python 3.6, so people can migrate to a new Python version.

We can reevaluate the numbers some time from now.

**Comment by tiangolo** [comment]
Yep! As @Kludex says, if possible, and if it's not being cumbersome to maintain, I would hold this for some time. :pray:

I assume there might already be "old" apps that can't migrate to 3.7 right away. If it gets annoying/cumbersome then yeah. But keeping support for 3.6 for a while might be appreciated by a few developers already stressed that their other libs don't support > 3.6 yet. :sweat_smile:

**Comment by Kludex** [comment]
`anyio` 4.0 will drop Python 3.6, jfyk.

**Comment by Kludex** [comment]
Trio dropped support for Python 3.6.

**Comment by Kludex** [comment]
Not important on the dependency tree, but `coverage` dropped support as well.

**Comment by Kludex** [comment]
`pytest` dropped support as well.

**Comment by Kludex** [comment]
Daily downloads from PyPI Download Stats about Starlette:

```
3.8: ~95k
3.7: ~78k
3.9: ~44k
3.10: ~16k
3.6: ~10k
```
3.18% of our users use Python 3.6.

Ref.: https://pypistats.org/packages/starlette (date of the info above is 29 March 2022)

Based on this, I'd like to go with this idea.

**Review by euri10 — APPROVED** [review]

**Review by br3ndonland — APPROVED** [review]
Thank you for the thoughtful, data-driven approach to this update. :chart_with_upwards_trend: :eyes: :thumbsup:

**Review by adriangb — APPROVED** [review]

File: README.md
Comment (Kludex) [inline README.md]:
```suggestion
Python 3.7+ (For Python 3.6 support, install version 0.19.1)
```

File: docs/index.md
Comment (Kludex) [inline docs/index.md]:
```suggestion
Python 3.7+ (For Python 3.6 support, install version 0.19.1)
```

**Pull request merged**
